enhancements/installer/cluster-api-based-installations.md

```diff
@@ -0,0 +1,485 @@
+its own infrastructure.
+
+The Cluster API manifests will be generated with `.spec` fields in the [manifest asset target][asset] with `.status`
+updated after cluster creation.
+
+#### Infrastructure Provisioning
+
+##### CAPI Provisioning
+
+The CAPI infrastructure manifests determine the infrastructure resources that will be provisioned by the CAPI controllers.
+Here is an example AWS infra cluster manifest from the proof-of-concept implementation:
+
+```shell
+$ cat cluster-api/02_infra-cluster.yaml
+apiVersion: v1beta2
+kind: AWSCluster
+metadata:
+  creationTimestamp: null
+  name: capi-example-kbdpg
+  namespace: openshift-cluster-api-guests
+spec:
+  bastion:
+    enabled: false
+  controlPlaneEndpoint:
+    host: ""
+    port: 0
+  controlPlaneLoadBalancer:
+    additionalListeners:
+    - port: 22623
+      protocol: TCP
+    crossZoneLoadBalancing: false
+    loadBalancerType: nlb
+    name: capi-example-kbdpg-ext
+    scheme: internet-facing
+  network:
+    additionalControlPlaneIngressRules:
+    - description: MCS traffic from cluster network
+      fromPort: 22623
+      protocol: tcp
+      sourceSecurityGroupRoles:
+      - node
+      - controlplane
+      toPort: 22623
+    - description: controller-manager
+      fromPort: 10257
+      protocol: tcp
+      sourceSecurityGroupRoles:
+      - controlplane
+      - node
+      toPort: 10257
+    - description: kube-scheduler
+      fromPort: 10259
+      protocol: tcp
+      sourceSecurityGroupRoles:
+      - controlplane
+      - node
+      toPort: 10259
+    cni:
+      cniIngressRules:
+      - description: ICMP
+        fromPort: -1
+        protocol: icmp
+        toPort: -1
+       # several rules omitted for brevity
+      - description: Service node ports (UDP)
+        fromPort: 30000
+        protocol: udp
+        toPort: 32767
+    vpc:
+      availabilityZoneSelection: Ordered
+      availabilityZoneUsageLimit: 3
+      cidrBlock: 10.0.0.0/16
+  region: us-east-2
+  s3Bucket:
+    name: openshift-bootstrap-data-capi-example-kbdpg
+    presignedURLDuration: 1h0m0s
+status:
+  networkStatus:
+    apiServerElb:
+      attributes: {}
+  ready: false
+```
+
+This manifest describes the network, load balancers, security group rules, etc. that will be used or provisioned for
+the target cluster. Control plane machines are provisioned by [AWSMachines][AWSMachines] & [CAPI Machines][CAPIMachines]
+manifests.
+
+##### Additional Infrastructure
+
+The Installer will generate any additional infrastructure that is needed but not handled by the CAPI provider, either
+because it is out of scope (e.g. IAM, DNS) or not adopted upstream (e.g. split-horizon load balancers). The Installer
+will provide hooks into the provisioning lifecycle that can be used to provision resources using direct SDK calls
+or other tooling.
+
+The AWS proof-of-concept implementation utilizes hooks defined in this interface:
+
+```go
+type CAPIInfraHelper interface {
```

Comment:
For increased flexibility, each function could be defined in a separate interface on the consuming (common) logic.

Reply:
The design suggested here is for each cloud platform to have its own implementation of the interface. Is that what you had in mind? Or can you elaborate a bit more on your suggestion.

Reply:
I think you mean something along the lines of
```
type CAPIInfraHelper interface {
  CAPIInfraPreProvisionHelper
  CAPIInfraControlPlaneAvailableHelper
}

type CAPIInfraPreProvisionHelper interface{
  PreProvision(in PreProvisionInput) error
}

type CAPIInfraControlPlaneAvailableHelper interface {
  ControlPlaneAvailable(in ControlPlaneAvailableInput) error
}
```
Then the implementation could just implement the subset of helpers that they need, assuming the installer code can correctly assert the the interfaces on whatever it is being passed in?

Reply:
That is what I had in mind, thank you. the `CAPIInfraHelper interface` becomes unnecessary one the other two are defined and individually asserted.